src/Features/Core/Portable/Diagnostics/EngineV2/DiagnosticIncrementalAnalyzer.StateManager.HostStates.cs

```diff
@@ -15,7 +15,19 @@ internal partial class DiagnosticIncrementalAnalyzer
         private partial class StateManager
         {
             public IEnumerable<StateSet> GetAllHostStateSets()
-                => _hostAnalyzerStateMap.Values.SelectMany(v => v.OrderedStateSets);
```

Comment:
@sharwell @ryzngard need your inputs here.

This was returning values for solution analyzers from all solution snapshots, instead of current solution, which led to duplicates being returned and subsequently an assert was firing at https://sourceroslyn.io/#Microsoft.CodeAnalysis.Features/Diagnostics/EngineV2/DiagnosticIncrementalAnalyzer.StateManager.cs,309 in integration tests.

This commit has one way to fix this issue. Another way to fix it would be to restore `_hostAnalyzerStateMap` back to only have the language name as the key as before, and introduce a separate field in the state manager for current solution state analyzer references. When asked to compute host state sets, we would first check if given project's solution state analyzer references are the same as ones stored onto this new field, otherwise we clear  `_hostAnalyzerStateMap` and recompute.

Reply:
I think strongly tying the `_hostAnalyzerStateMap` to requirements of uniqueness makes sense in the way it's already done. It makes it very clear what the inputs used for the current state map are without having to look at multiple fields. I know it does add overhead of passing more values, but I think overall I still like the current approach.